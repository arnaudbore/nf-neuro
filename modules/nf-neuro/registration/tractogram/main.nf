process REGISTRATION_TRACTOGRAM {
    tag "$meta.id"
    label 'process_single'

    container "scilus/scilus:2.2.0"

    input:
    tuple val(meta), path(anat), path(affine), path(tractogram), path(reference), path(deformation)

    output:
    tuple val(meta), path("*.{trk,tck}") , emit: tractogram
    path "versions.yml"                  , emit: versions

    when:
    task.ext.when == null || task.ext.when

    script:
    def prefix = task.ext.prefix ?: "${meta.id}"
    def suffix = task.ext.suffix ? "_${task.ext.suffix}" : ""
    reference = "$reference" ? "--reference $reference" : ""
    def in_deformation = "$deformation" ? "--in_deformation $deformation" : ""

    def inverse = task.ext.inverse ? "--inverse" : ""
    def reverse_operation = task.ext.reverse_operation ? "--reverse_operation" : ""

    def invalid_management = task.ext.invalid_streamlines ?: "cut"
    def cut_invalid = invalid_management == "cut" ? "--cut_invalid" : ""
    def remove_single_point = task.ext.remove_single_point ? "--remove_single_point" : ""
    def remove_overlapping_points = task.ext.remove_overlapping_points ? "--remove_overlapping_points" : ""
    def threshold = task.ext.threshold ? "--threshold " + task.ext.threshold : ""
    def no_empty = task.ext.no_empty ? "--no_empty" : ""

    """
    affine=$affine
    if [[ "$affine" == *.txt ]]; then
        ConvertTransformFile 3 $affine affine.mat --convertToAffineType \
            && affine="affine.mat" \
            || echo "TXT affine transform file conversion failed, using original file."
    fi

    for tractogram in ${tractogram}; do
        ext=\${tractogram#*.}
        bname=\$(basename \${tractogram} .\${ext} | sed 's/${prefix}_\\+//')
        name=${prefix}_\${bname}${suffix}.\${ext}

        scil_tractogram_apply_transform \$tractogram $anat \$affine \$name \
            $in_deformation \
            $inverse \
            $reverse_operation \
            $reference \
            --keep_invalid -f

        if [[ "$invalid_management" == "keep" ]]; then
            echo "Skip invalid streamline detection: \$name"
            continue
        fi

        scil_tractogram_remove_invalid \$name \$name \
            $cut_invalid\
            $remove_single_point\
            $remove_overlapping_points\
            $threshold\
            $no_empty\
            -f
    done

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        ants: \$(antsRegistration --version | grep "Version" | sed -E 's/.*v([0-9.a-zA-Z-]+).*/\\1/')
        scilpy: \$(uv pip -q -n list | grep scilpy | tr -s ' ' | cut -d' ' -f2)
    END_VERSIONS
    """

    stub:
    def prefix = task.ext.prefix ?: "${meta.id}"
    def suffix = task.ext.suffix ? "_${task.ext.suffix}" : ""
    """
    scil_tractogram_apply_transform -h
    scil_tractogram_remove_invalid -h

    for tractogram in ${tractogram}; do
        ext=\${tractogram#*.}
        bname=\$(basename \${tractogram} .\${ext} | sed 's/${prefix}_\\+//')
        name=${prefix}_\${bname}${suffix}.\${ext}
        touch \$name
    done

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        ants: \$(antsRegistration --version | grep "Version" | sed -E 's/.*v([0-9.a-zA-Z-]+).*/\\1/')
        scilpy: \$(uv pip -q -n list | grep scilpy | tr -s ' ' | cut -d' ' -f2)
    END_VERSIONS
    """
}
