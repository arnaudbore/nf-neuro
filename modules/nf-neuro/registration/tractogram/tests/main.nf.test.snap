{
    "registration - tractogram - keep invalid": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test_bundle_2.trk:md5,824bcbf796cbce3e1c93e39d1c98dcf9"
                    ]
                ],
                "1": [
                    "versions.yml:md5,452295f58ea9ab6aa25264de21b111c1"
                ],
                "tractogram": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test_bundle_2.trk:md5,824bcbf796cbce3e1c93e39d1c98dcf9"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,452295f58ea9ab6aa25264de21b111c1"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.0",
            "nextflow": "25.04.8"
        },
        "timestamp": "2025-10-16T15:24:45.131144667"
    },
    "registration - tractogram_bundles": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        [
                            "test_bundle_0.trk:md5,6e14cc02b66d12d5dde0a0701918385a",
                            "test_bundle_1.trk:md5,ef5a759144ab7e8d2f5abf30a463a152",
                            "test_bundle_2.trk:md5,824bcbf796cbce3e1c93e39d1c98dcf9",
                            "test_bundle_3.trk:md5,8877161b9aebaa6b12a4d7a8f8410407",
                            "test_bundle_4.trk:md5,6c8658b946154896952ca89e88fcf29f",
                            "test_bundle_5.trk:md5,fa058f93d9bfe4bed2ff027e9acd5682",
                            "test_bundle_6.trk:md5,125e50561b4b0c5fd8ee23867c28bc51"
                        ]
                    ]
                ],
                "1": [
                    "versions.yml:md5,452295f58ea9ab6aa25264de21b111c1"
                ],
                "tractogram": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        [
                            "test_bundle_0.trk:md5,6e14cc02b66d12d5dde0a0701918385a",
                            "test_bundle_1.trk:md5,ef5a759144ab7e8d2f5abf30a463a152",
                            "test_bundle_2.trk:md5,824bcbf796cbce3e1c93e39d1c98dcf9",
                            "test_bundle_3.trk:md5,8877161b9aebaa6b12a4d7a8f8410407",
                            "test_bundle_4.trk:md5,6c8658b946154896952ca89e88fcf29f",
                            "test_bundle_5.trk:md5,fa058f93d9bfe4bed2ff027e9acd5682",
                            "test_bundle_6.trk:md5,125e50561b4b0c5fd8ee23867c28bc51"
                        ]
                    ]
                ],
                "versions": [
                    "versions.yml:md5,452295f58ea9ab6aa25264de21b111c1"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.0",
            "nextflow": "25.04.8"
        },
        "timestamp": "2025-10-16T15:23:46.56008019"
    },
    "registration - tractogram - stub-run": {
        "content": [
            [
                "versions.yml:md5,452295f58ea9ab6aa25264de21b111c1"
            ]
        ],
        "meta": {
            "nf-test": "0.9.0",
            "nextflow": "25.04.7"
        },
        "timestamp": "2025-10-02T20:59:23.756300025"
    },
    "registration - tractogram - custom suffix": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test_bundle_2_mni.trk:md5,824bcbf796cbce3e1c93e39d1c98dcf9"
                    ]
                ],
                "1": [
                    "versions.yml:md5,452295f58ea9ab6aa25264de21b111c1"
                ],
                "tractogram": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test_bundle_2_mni.trk:md5,824bcbf796cbce3e1c93e39d1c98dcf9"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,452295f58ea9ab6aa25264de21b111c1"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.0",
            "nextflow": "25.04.8"
        },
        "timestamp": "2025-10-16T15:24:16.677031753"
    }
}